k-distribution/src/main/scripts/lib/py-kframework/kast.py

```diff
@@ -0,0 +1,228 @@
+    return { "node" : "KToken", "token": token, "sort": sort}
+
+def isKToken(k):
+    return k["node"] == "KToken"
+
+def KAtt(key, value):
+    return {"node": "KAtt", "key": key, "value": value}
+
+def isKAtt(k):
+    return k["node"] == "KAtt"
+
+def KRule(rule, requires = None, ensures = None, atts = None):
+    return { "node": "KRule", "rule": rule, "requires": requires, "ensures": ensures, "atts": atts }
+
+def isKRule(k):
+    return k["node"] == "KRule"
+
+def KImport(kimport):
+    return { "node": "KImport", "import": kimport }
+
+def isKImport(k):
+    return k["node"] == "KImport"
+
+def KModule(name, imports, rules):
+    return { "node": "KModule", "name": name, "imports": imports, "rules": rules }
+
+def isKModule(k):
+    return k["node"] == "KModule"
+
+def KRequire(krequire):
+    return { "node": "KRequire", "require": krequire }
+
+def isKRequire(k):
+    return k["node"] == "KRequire"
+
+def KDefinition(name, requires, modules):
+    return { "node": "KDefinition", "name": name, "requires": requires, "modules": modules }
+
+def isKDefinition(k):
+    return k["node"] == "KDefinition"
+
+def isCellKLabel(label):
+    return label[0] == "<" and label[-1] == ">"
+
+def addAttributes(kast, atts):
+    if isKRule(kast):
+        newAtts = []
+        if kast["atts"] is None:
+            newAtts = atts
+        else:
+            newAtts.extend(kast["atts"])
+        return KRule(kast["rule"], requires = kast["requires"], ensures = kast["ensures"], atts = newAtts)
+    else:
+        notif("Don't know how to add attributes to KAST!")
+        print(kast)
+        sys.exit(1)
+
+klabelRewrite = "#KRewrite"
+klabelCells   = "#KCells"
+klabelEmptyK  = "#EmptyK"
+
+ktokenDots = KToken("...", "K")
+
+def paren(printer):
+    return (lambda *args: "( " + printer(*args) + " )")
+
+def binOpStr(symbol):
+    return (lambda a1, a2: a1 + " " + symbol + " " + a2)
+
+def appliedLabelStr(symbol):
+    return (lambda *args: symbol + "( " + " , ".join(args) + " )")
+
+def constLabel(symbol):
+    return (lambda: symbol)
+
+def assocWithUnit(assocJoin, unit):
+    def _assocWithUnit(*args):
+        newArgs = [ arg for arg in args if arg != unit ]
+        return assocJoin.join(newArgs)
+    return _assocWithUnit
+
+def underbarUnparsing(symbol):
+    splitSymbol = symbol.split("_")
+    numArgs = len([symb for symb in splitSymbol if symb == ""])
+    def _underbarUnparsing(*args):
+        result = []
+        i = 0
+        for symb in splitSymbol:
+            if symb != "":
+                result.append(symb)
+            if i < len(args):
+                result.append(args[i])
+                i += 1
+        return " ".join(result)
+    return _underbarUnparsing
+
+K_builtin_labels = { klabelRewrite : binOpStr("=>")
+                   , klabelCells   : (lambda *args: "\n".join(args))
+                   , klabelEmptyK  : (lambda : ".")
+                   }
+
+BOOL_expressions = { "_andBool__BOOL" : binOpStr("\nandBool")
```

Comment:
this is not robust. relying on the klabels generated for productions that don't have the klabel and symbol attributes is going to be fragile and in particular it will break in the llvm and haskell backends. if you want to refer to all of these labels explicitly, let's give them explicit klabels and the symbol attribute.

Reply:
I've added explicit `klabel` and `symbol` attributes for many parts of `domains.k`, at least all the parts that `pyk` understands for now.